### ISSUE TITLE
Hosting in a .Net Core 2.1 Windows Service on Windows Server 2016

### ISSUE THREAD

**Issue by mmaguirehub** [aspnet/Docs#7373]
I've followed the instructions here to the letter and my .Net Core 2.1 Windows Service works perfectly when I create and start the service on my Windows 10 Pro machine.

The problem arises when I deploy on to our Windows Server 2016 Datacenter OS. It creates the service as expected, but when I go to start the service I immediately get the following error:

The service did not respond to the start or control request in a timely fashion.

To emphasize, the service instantly fails to start with the above error.

If I try to run the exe itself I get the following error:

Error:
An assembly specified in the application dependencies manifest (TestService.deps.json) was not found:
package: 'Microsoft.AspNetCore.Hosting.WindowsServices', version: '2.1.1'
path: 'lib/netstandard2.0/Microsoft.AspNetCore.Hosting.WindowsServices.dll'

I suspect the two error are interlinked somehow.

I'm fairly convinced that the code is not being hit in Program.Main as I've added logging in the first line of code to be ran but nothing happens. I've tried adding the ServicesPipeTimeout registry entry to extend the time given for the service to start but I still get an instant fail. I've ensured that the server has the required versions of .Net Core installed and they are:

Microsoft. NET Core 2.1.1 - Windows Server Hosting - v2.1.13007.0
Microsoft .NET Core Runtime - 2.1.1 (x64) - 2.1.1.26606
I've also ensured the IIS is enabled on the system.

I've been struggling with this for 2 days and cannot find the answers to resolve these issues anywhere.

Can this document be updated with instructions on how to deploy to the server to ensure all the pre-requisites are covered so that these issues don't occur?

I'd really like an answer to this, thanks in advance :)

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: b0bd8cb8-7e28-8c4d-a550-3aea8f3ab4dd
* Version Independent ID: 4b60d24e-ef05-15e5-10b4-483094081bf3
* Content: [Host ASP.NET Core in a Windows Service](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/windows-service?view=aspnetcore-2.1)
* Content Source: [aspnetcore/host-and-deploy/windows-service.md](https://github.com/aspnet/Docs/blob/master/aspnetcore/host-and-deploy/windows-service.md)
* Product: **aspnet-core**
* GitHub Login: @guardrex
* Microsoft Alias: **tdykstra**

**Comment by guardrex** [comment]
Hello @mmaguirehub ... Do you have a package reference for `Microsoft.AspNetCore.Hosting.WindowsServices` in the project file?

**Comment by mmaguirehub** [comment]
Hi @guardrex, yes the package reference is there, here is the csproj:

```
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>netcoreapp2.1</TargetFramework>
    <RuntimeIdentifier>win7-x64</RuntimeIdentifier>
  </PropertyGroup>

  <ItemGroup>
    <Folder Include="wwwroot\" />
  </ItemGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.AspNetCore.App" />
    <PackageReference Include="Microsoft.AspNetCore.Hosting.WindowsServices" Version="2.1.1" />
  </ItemGroup>

</Project>
```
As you can see the version is set correctly at 2.1.1.

**Comment by guardrex** [comment]
You're publishing the project and creating the service from the `publish` folder? `bin\release\netcoreapp2.1\publish`

btw- You won't need the `Folder Include`, as using the `Microsoft.NET.Sdk.Web` SDK has a target to move those assets on publish.

**Comment by guardrex** [comment]
Your SDK is 2.1.300? `dotnet --version`

**Comment by mmaguirehub** [comment]
Thanks @guardrex, I wasn't creating the service from the `Publish` folder.  I've just tried that and it now works which is a massive relief.  I'll now remove the `Folder Include`.

The SDK is 2.1.301.

**Comment by guardrex** [comment]
Ok, so the problem is in Step 4 of the topic. I'm going to make an enhancement there to make sure that readers know. Although Step 3 calls for publishing, Step 4 doesn't say to create the service from the *publish* folder.

I'll take care of it. Sorry u had so much trouble. We don't find out what's wrong with topics until someone (*you!*) runs into a problem.

**Comment by mmaguirehub** [comment]
@guardrex How do I make it so that it only those dependencies required are published, it currently looks like everything is output and its currently 90Mb.  I'm sure I don't need all those files

**Comment by guardrex** [comment]
The level of trimming is rather modest I believe ... I'm just basing that on some remarks that `@DamianEdwards` has made about dependency trimming in general. The sample app has the same result. The *publish* folder comes in at 83.7 MB. I'm afraid there's nothing that can be done that I'm aware of to further trim.

I'd like to run a quick test tho. I have the prior set of specific packages for the sample app here. Let me toss those into the sample, publish it, and see if it shrinks the app at all over using the `App` package.

Be back in a few minutes with the answer ...

**Comment by guardrex** [comment]
Ah ... turns out that I'm working on a topic on an *old branch* that has the prior version of the sample. That version calls for the packages explicitly ...

```json
<PackageReference Include="Microsoft.AspNetCore" Version="2.1.0" />
<PackageReference Include="Microsoft.AspNetCore.CookiePolicy" Version="2.1.0" />
<PackageReference Include="Microsoft.AspNetCore.Hosting.WindowsServices" Version="2.1.0" />
<PackageReference Include="Microsoft.AspNetCore.HttpsPolicy" Version="2.1.0" />
<PackageReference Include="Microsoft.AspNetCore.Mvc" Version="2.1.0" />
<PackageReference Include="Microsoft.AspNetCore.StaticFiles" Version="2.1.0" />
<PackageReference Include="Microsoft.Extensions.Logging.Debug" Version="2.1.0" />
```
... and the result is 83.7 MB. So yeah ... unfortunately, that's producing the size I see.

You could ask as a general question on the aspnet/Home issues. One of the engineers there might comment on the state of package trimming and possibly even what they plan in the future.

https://github.com/aspnet/Home/issues

**Comment by mmaguirehub** [comment]
Thanks @gaurdrex for your help on this.
I'll do as you say and ask that general question on aspnet/home issues.

**Comment by bikeoid** [comment]
FYI - not sure if this is related, I see this error also even though I am referencing results of the publish folder.
 I get this service timeout error only when installing in a path that uses a space in the path name (such as "C:\Program Files").  I can solve my issue by installing to a path without spaces.

  I don't see the error when running interactively with the --console argument.

 I got the idea to try a path without spaces by referencing an issue Peter Kottas saw - https://github.com/PeterKottas/DotNetCore.WindowsService/issues/13

**Comment by bikeoid** [comment]
Thanks - I finally tracked my problem down and it was an error in my  appsettings.json - I had grabbed the wrong appsettings.json for that test.   It had an unescaped '\' in the db connection string, which made the JSON parse fail.   This was further complicated because it happens before the service startup handshake completes, which apparently doesn't log the error to the Windows Event Viewer.

So now it works as a service in paths with spaces when I use the correct appsettings.json
